src/archive/src/main.rs

```diff
@@ -0,0 +1,438 @@
+                config.ii_canister
+            ))
+        }
+    });
+    let idx = with_log(|log| {
+        log.append(entry.as_ref())
+            .expect("failed to append log entry")
+    });
+
+    with_anchor_index_mut(|index| {
+        let key = AnchorIndexKey {
+            anchor,
+            timestamp,
+            log_index: idx as u64,
+        };
+
+        // the only way this expect can trigger is when the key size is wrong.
+        // On other failures (e.g. no more stable memory available) the underlying StableBTreeMap will panic directly.
+        index.insert(key, ()).expect("bug: key size mismatch");
+    })
+}
+
+#[query]
+fn get_entries(index: Option<u64>, limit: Option<u16>) -> Entries {
+    let num_entries = limit_or_default(limit);
+
+    with_log(|log| {
+        let length = log.len();
+        let start_idx = match index {
+            None => length.saturating_sub(num_entries),
+            Some(idx) => idx as usize,
+        };
+
+        let next_idx = if start_idx + num_entries < length {
+            Some((start_idx + num_entries) as u64)
+        } else {
+            None
+        };
+
+        let mut entries = Vec::with_capacity(num_entries);
+        for idx in start_idx..start_idx + num_entries {
+            let entry = match log.get(idx) {
+                None => break,
+                Some(entry) => entry,
+            };
+            entries.push(Some(
+                candid::decode_one(&entry).expect("failed to decode log entry"),
+            ))
+        }
+        Entries { entries, next_idx }
+    })
+}
+
+#[query]
+fn get_anchor_entries(anchor: Anchor, cursor: Option<Cursor>, limit: Option<u16>) -> AnchorEntries {
+    let num_entries = limit_or_default(limit);
+
+    with_anchor_index_mut(|index| {
+        let iterator = match cursor {
+            None => index.range(anchor.to_le_bytes().to_vec(), None),
+            Some(Cursor::NextToken { next_token }) => index.range(next_token.into_vec(), None),
+            Some(Cursor::Timestamp { timestamp }) => index.range(
+                anchor.to_le_bytes().to_vec(),
+                Some(timestamp.to_le_bytes().to_vec()),
+            ),
+        };
+
+        with_log(|log| {
+            let mut entries: Vec<(AnchorIndexKey, Vec<u8>)> = iterator
+                .take(num_entries + 1) // take one too many to extract the cursor
+                .map(|(anchor_key, _)| {
+                    let entry = log
+                        .get(anchor_key.log_index as usize)
+                        .expect("bug: index to non-existing entry");
+                    (anchor_key, entry)
+                })
+                .collect();
+
+            let cursor = if entries.len() > num_entries {
+                entries.pop().map(|(key, _)| Cursor::NextToken {
+                    next_token: ByteBuf::from(key.to_bytes()),
+                })
+            } else {
+                None
+            };
```

Comment:
Wouldn't it be simpler to first extract `num_entries` and then read `iterator.nth(num_entries + 1)`? Looks like the index [only implements `next()`](https://docs.rs/ic-stable-structures/latest/src/ic_stable_structures/btreemap/iter.rs.html#89), so I'm guessing this is an optimization to avoid iterating twice?

Reply:
The problem is that `collect` consumes the iterator and iterating twice seems wasteful. Are you ok with leaving this as is? Or have a suggestion to structure this more nicely?

Reply:
> iterating twice seems wasteful

Ok, that's what I wanted to clarify.

I'm completely fine with it as is! I wanted to make sure I understood what was happening, so I think it's at least worth a comment.

Another approach would be to use `.next()` directly and explicitly; might be a little more verbose but the intent would be clearer. Happy with current implementation though.